torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -377,6 +378,318 @@ bool IsBlockReturnTypeSame(Node* n) {
+    }
+  }
+  if (inputTensorValues.size() < n->inputs().size()) {
+    return c10::nullopt;
+  }
+  try {
+    return onnx_constant_fold::runTorchBackendForOnnx(
+        n, inputTensorValues, opset_version);
+  } catch (const std::exception& ex) {
+    TORCH_WARN(
+        "Constant folding in symbolic shape inference fails: ", ex.what());
+    return c10::nullopt;
+  }
+}
+
+// When the Reshape node's two inputs are constant, compute the output shape.
+// The reshape value 0 and -1 are converted to the real value explicitly.
+std::vector<int64_t> ComputeShapeFromReshape(
+    const std::vector<int64_t>& input_shape,
+    const std::vector<int64_t>& reshape) {
+  TORCH_INTERNAL_ASSERT(
+      input_shape.size() > 0 || reshape.size() > 0,
+      "Reshape node should have at least one input size > 0 when constant folding.");
+  // Case: reshape.size() == 0
+  // %22 : int[] = prim::Constant[value=annotate(List[int], [])]()
+  // %15 : Float(requires_grad=0, device=cpu) = aten::view(%1, %22)
+  if (reshape.size() == 0) {
+    return input_shape;
+  }
+  if (input_shape.size() == 0) {
+    return reshape;
+  }
+  auto reshape_has_zero = false;
+  auto reshape_size = static_cast<int>(reshape.size());
+  for (auto i = 0; i < reshape_size; i++) {
+    if (reshape[i] == 0) {
+      reshape_has_zero = true;
+      break;
+    }
+  }
+  int minus_one_pos = -1;
+  auto input_shape_size = static_cast<int>(input_shape.size());
+  for (auto i = 0; i < input_shape_size; i++) {
+    if (reshape[i] == -1) {
+      minus_one_pos = i;
+      break;
+    }
+  }
+  if (!reshape_has_zero && minus_one_pos == -1) {
+    return reshape;
+  }
+  std::vector<int64_t> final_shape;
+  // shape_ratio is used to calculate the real value of -1.
+  // it is updated on each dimension to avoid overflow.
+  double shape_ratio = 1.0;
+  for (auto i = 0; i < input_shape_size; i++) {
+    if (i < minus_one_pos) {
+      if (reshape[i] != 0) {
+        shape_ratio *= static_cast<double>(input_shape[i]) / reshape[i];
+      }
+    } else if (minus_one_pos > -1) {
+      if (i < input_shape_size - reshape_size + minus_one_pos + 1) {
+        shape_ratio *= static_cast<double>(input_shape[i]);
+      } else {
+        auto reshape_idx = i - input_shape_size + reshape_size;
+        if (reshape[reshape_idx] != 0) {
+          shape_ratio *=
+              static_cast<double>(input_shape[i]) / reshape[reshape_idx];
+        }
+      }
+    }
+  }
+
+  for (auto i = 0; i < minus_one_pos; i++) {
+    int64_t cur_shape = reshape[i] == 0 ? input_shape[i] : reshape[i];
+    final_shape.push_back(cur_shape);
+  }
+  final_shape.push_back(static_cast<int64_t>(std::round(shape_ratio)));
+  for (auto i = minus_one_pos + 1; i < reshape_size; i++) {
+    int64_t cur_shape = reshape[i] == 0
+        ? input_shape[i + input_shape_size - reshape_size]
+        : reshape[i];
+    final_shape.push_back(cur_shape);
+  }
+  return final_shape;
+}
+
+void UpdateRank(Value* value, size_t rank) {
```

Comment:
is it safe to assume value already has tensortype?
and is it safe to assume the shape can be overwritten by the new symbolicShape?

Reply:
Add `if (TensorTypePtr value_type = value->type()->cast<TensorType>())`